File: java/AlgorithmAndDataStructure/src/main/java/algorithms/Numeric.java
package algorithms;

import functionObjects.BinaryOp;
import functionObjects.Itoa;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;

/**
 * Created by lzx on 15-2-22.
 */
public class Numeric {
	public static <T> void itoa(List<T> a, Itoa<T> it) {
		int size = a.size();
		for (int i = 0; i < size; i++) {
			a.set(i, it.value());
			it.increase();
		}
	}

	public static <T> T accumulate(Collection<T> a, T init, BinaryOp<T> op) {
		for (T o : a)
			init = op.call(init, o);
		return init;
	}

	public static <T> void innerProduct(Collection<T> a, Collection<T> b,
	                                    Collection<T> out, BinaryOp<T> op) {
		Iterator<T> ia = a.iterator();
		Iterator<T> ib = b.iterator();
		while (ia.hasNext() && ib.hasNext())
			out.add(op.call(ia.next(), ib.next()));
	}

	// should we rename it to adjacentOp?
	public static <T> void adjacentDifference(Collection<T> a, Collection<T> out,
	                                          BinaryOp<T> op) {
		Iterator<T> i = a.iterator();
		if (!i.hasNext())
			return;
		T prev = i.next();
		out.add(prev);
		while (i.hasNext()) {
			T cur = i.next();
			out.add(op.call(prev, cur));
			prev = cur;
		}
	}

	// call it partialReduce better
	public static <T> void partialSum(Collection<T> a, Collection<T> out,
	                                  BinaryOp<T> op) {
		Iterator<T> i = a.iterator();
		if (!i.hasNext())
			return;
		T sum = i.next();
		out.add(sum);
		while (i.hasNext()) {
			sum = op.call(sum, i.next());
			out.add(sum);
		}
	}
}

File: typescript/src/numeric.ts
let itoa = (init: number): ()=>number => {
    let value = init
    return (): number => { return value++ }
}

let accumulate = <T>(ary: T[], init: T, func: (a: T, b: T) => T): T => {
    return ary.reduce(func, init)
}

// assume ary2.length >= ary1.length
let innerProduction = (ary1: number[], ary2: number[], init: number): number => {
    return ary1.reduce((prev: number, cur: number, idx: number): number => {
        return prev + cur * ary2[idx]
    }, init)
}

let adjacentDifference = (ary: number[]): number[] => {
    if (ary.length == 0) {
        return new Array<number>()
    }
    let res: number[] = [ary[0]]
    for (let i = 1; i < ary.length; i++) {
        res.push(ary[i] - ary[i - 1])
    }
    return res
}

let partialSum = (ary: number[]): number[] => {
    if (ary.length == 0) {
        return new Array<number>()
    }
    let res: number[] = [ary[0]]
    ary.reduce((prev: number, cur: number): number => {
        res.push(prev + cur)
        return prev + cur
    })
    return res
}

export {
    itoa,
    accumulate,
    innerProduction,
    adjacentDifference,
    partialSum,
}
